metrics.go

```diff
@@ -0,0 +1,180 @@
+package tcp
+
+import (
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/marten-seemann/tcp"
+	"github.com/mikioh/tcpinfo"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	newConns    *prometheus.CounterVec
+	closedConns *prometheus.CounterVec
+)
+
+var collector *aggregatingCollector
+
+func init() {
+	collector = newAggregatingCollector()
+	prometheus.MustRegister(collector)
+
+	const direction = "direction"
+
+	newConns = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "tcp_connections_new_total",
+			Help: "TCP new connections",
+		},
+		[]string{direction},
+	)
+	prometheus.MustRegister(newConns)
+	closedConns = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "tcp_connections_closed_total",
+			Help: "TCP connections closed",
+		},
+		[]string{direction},
+	)
+	prometheus.MustRegister(closedConns)
+}
+
+type aggregatingCollector struct {
+	mutex sync.Mutex
+
+	highestID uint64
+	conns     map[uint64] /* conn ID */ *tracingConn
+}
+
+var _ prometheus.Collector = &aggregatingCollector{}
+
+func newAggregatingCollector() *aggregatingCollector {
+	return &aggregatingCollector{conns: make(map[uint64]*tracingConn)}
+}
+
+func (c *aggregatingCollector) AddConn(t *tracingConn) uint64 {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.highestID++
+	c.conns[c.highestID] = t
+	return c.highestID
+}
+
+func (c *aggregatingCollector) removeConn(id uint64) {
+	delete(c.conns, id)
+}
+
+func (c *aggregatingCollector) newRTTHistogram() prometheus.Summary {
+	return prometheus.NewHistogram(prometheus.HistogramOpts{
+		Name:    "tcp_rtt",
+		Help:    "RTT",
+		Buckets: prometheus.ExponentialBuckets(0.001, 1.25, 40), // 1ms to ~6000ms
+	})
+}
+
+func (c *aggregatingCollector) newConnectionDurationHistogram() prometheus.Summary {
+	return prometheus.NewHistogram(prometheus.HistogramOpts{
+		Name:    "tcp_connection_duration",
+		Help:    "Connection Duration",
+		Buckets: prometheus.ExponentialBuckets(1, 1.5, 40), // 1s to ~12 weeks
+	})
+}
+
+func (c *aggregatingCollector) Describe(descs chan<- *prometheus.Desc) {
+	descs <- c.newRTTHistogram().Desc()
+	descs <- c.newConnectionDurationHistogram().Desc()
+}
+
+func (c *aggregatingCollector) Collect(metrics chan<- prometheus.Metric) {
+	now := time.Now()
+	rtts := c.newRTTHistogram()
+	connDurations := c.newConnectionDurationHistogram()
```

Comment:
Why create a new Histogram on each collect? After all Prometheus histograms are designed to work over time, i.e. live for the lifetime of the process. I feel like we have discussed this in the past, thus I am sorry in case I am missing something.

Reply:
I thought I had fixed that, but apparently forgot to include it in the PR. Sorry!